Add missing calls to raxStop

hwware opened the pull request.

oranagra commented:
i now see that there are lots of places that forget to call `raxStop`, eg. `rdbSaveStreamPEL`, or even don't bother doing that at all, e.g. `handleBlockedClientsTimeout`.
lucky this usually doesn't cause any leak since the rax stack is only allocated for deep trees, but this makes it much more dangerous (a leak that is not detected in the valgrind test suite, but will happen on big production deployments).

@hwware may i ask how did you find that?
do you want to take this opportunity and go over the other places in this PR or a new one?

hwware commented:
@oranagra hello Oran, sure I can definitely take look other places which doesn't call raxStop and I will expand this pr. I found this when I was reading the source code for defrag.c (this part I am not very familiar with the detailed implementation). I don't think valgrind can easily detect all of these leak issue since some of them maybe in edge cases and our test may doesn't cover all these cases.

oranagra commented:
it bothers me that there are so many places where raxStop was missing (as if done on purpose), and that it didn't cause any actual harm so far.

as far as i can tell, since RAX_STACK_STATIC_ITEMS is 32 and RAX_ITER_STATIC_LEN is 128, it means that trees with keys that are guaranteed to always be shorter than 32 bytes, will never leak, even if not calling raxStop at all.
however, that's obviously not a reason not to call raxStop.
@guybe7 can you please acknowledge that's right?

so the two examples i gave above, `rdbSaveStreamPEL` (and the defrag.c "leak") will never leak since the keys are stream IDs which are always 16 bytes, similarly `handleBlockedClientsTimeout` will never leak since the key names are also always 16 bytes (CLIENT_ST_KEYLEN). 

@hwware while you're reviewing all the places where raxStop was missing (and fixing them), can you please provide a list of the collections they're iterating and which of these (if any) have indeed keys that can be 32 bytes or longer.

thanks a lot.

hwware commented:
@oranagra Hello Oran, thank you for your information. Here are the places I found raxStop was missing:

https://github.com/redis/redis/blob/unstable/src/timeout.c#L136

This should be safe, since the Key is a 8bytes mstime plus 8 bytes client id, total 16 bytes..

https://github.com/redis/redis/blob/unstable/src/defrag.c#L643
this is about this pr, and you are right, stream id are always 16 bytes so should be safe here.

https://github.com/redis/redis/blob/unstable/src/rdb.c#L686
Here we used stream id as the key in the pending entry list, 16 bytes should be safe here too.

https://github.com/redis/redis/blob/unstable/src/rdb.c#L722
here we may have problem, since we using consumer name as the key, which is user specified, AFAIK this can be larger than 32 bytes.

In rdbSaveObject function, we have lots of places missing raxStop call, in line 
https://github.com/redis/redis/blob/unstable/src/rdb.c#L915 
And line 917, we saving stream id here as the key, so should not cause leaking. 
But in line https://github.com/redis/redis/blob/unstable/src/rdb.c#L949 and rest, since cgroups radix tree we are saving consumer group name,  AFAIK this can also be larger than 32 bytes.

https://github.com/redis/redis/blob/unstable/src/aof.c#L1240
All the places missing raxStop below seems dangerous too, since here we are iterating from the cgroups and consumers, as the same reason if the cgroup name or consumer name is larger than 32 bytes it may introduce leaking.

Even though we did not get leaking problem in some of these places, I still think it is a good idea to fix it, this can avoid future headache for us, if we changed something such as making stream id length configurable in the future, it will make this unsafe.

Please take look this and let me know if you agree, I will work on a fix for all these places. thanks.

oranagra commented:
Thank you @hwware
Yes, we should certainly fix all of them. The interface of rax.h requires to call the raxStop, and the details about when it leaks and when it doesn't may change some day. 
So please go ahead and update the PR. 

I just wanted to know how come we don't notice it till now and if it poses any real problem for past releases. 
Luckily, all the places that have a real leak are code paths that only run inside a fork child or at shutdown, so these leaks can in theory accumulate to a lot of memory, but will not remain for long time. 
Also, I hope not many use group and consumer names that long..

hwware commented:
@oranagra Hi Oran, I have finished editing these changes for fixing missing raxStop call. Please review this again, thanks!

oranagra commented on src/aof.c:
```diff
@@ -1244,12 +1244,30 @@ int rewriteStreamObject(rio *r, robj *key, robj *o) {
+            if (rioWriteBulkStreamID(r,&group->last_id) == 0) {
+                raxStop(&ri);
+                return 0;
+            }
```
I don't like how the error handling changed that code from being short and easy to read to be bloated with error handling code.

i see 3 options:
1. use a single `if` with many `||` so we'll end up with just 6 lines of writes, followed by 3 lines of error handling.
2. use 2 `goto` statements
3. remove the newlines and do `if () {raxStop(); return 0;}` all in one line
4. keep it like you did.

i feel 1 is the best one, let me know if you see other options.

oranagra commented on src/rdb.c:
```diff
@@ -912,30 +930,40 @@ ssize_t rdbSaveObject(rio *rdb, robj *o, robj *key) {
          * this easily later, since our macro nodes should be checked for
          * number of items: not a great CPU / space tradeoff. */
-        if ((n = rdbSaveLen(rdb,s->length)) == -1) return -1;
+        if ((n = rdbSaveLen(rdb,s->length)) == -1) 
```
personally, i do indeed dislike seeing statements that change the execution flow as part of the `if` line.
except for cases where a certain pattern is repeated many times (especially line after line like in the case we've seen in aof.c, where i'd rather have them in one line).

but anyway, another thing i don't like is changing code without a good reason (makes it harder to find things in the blame log).

i'd rather you revert the ones which you didn't really change.

hwware commented:
@oranagra Hi Oran, based on your comments, I changed my code in two parts, 1. I used a single or statement for the condition check in aof.c. 2. I removed all the coding style fix and only include the functional changes. please review again. Thanks!

oranagra commented:
thanks @hwware. edited the commit message and merged.
